nucliadb/nucliadb/ingest/consumer/materializer.py

```diff
@@ -0,0 +1,89 @@
+# Copyright (C) 2021 Bosutech XXI S.L.
+#
+# nucliadb is offered under the AGPL v3.0 and as commercial software.
+# For commercial licensing, contact us at info@nuclia.com.
+#
+# AGPL:
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+#
+
+import logging
+import uuid
+from functools import partial
+
+from nucliadb.common.cluster.utils import get_shard_manager
+from nucliadb.common.datamanagers.resources import ResourcesDataManager
+from nucliadb.common.maindb.driver import Driver
+from nucliadb_protos import writer_pb2
+from nucliadb_utils import const
+from nucliadb_utils.cache.pubsub import PubSubDriver
+from nucliadb_utils.storages.storage import Storage
+
+from .utils import DelayedTaskHandler
+
+logger = logging.getLogger(__name__)
+
+
+class MaterializerHandler:
+    """
+    The purpose of this component is to materialize data that is
+    expensive to calculate at read time and is okay to be stale.
+    """
+
+    subscription_id: str
+
+    def __init__(
+        self,
+        *,
+        driver: Driver,
+        pubsub: PubSubDriver,
+        storage: Storage,
+        check_delay: float = 30.0,
+    ):
+        self.resources_data_manager = ResourcesDataManager(driver, storage)
+        self.pubsub = pubsub
+        self.shard_manager = get_shard_manager()
+        self.task_handler = DelayedTaskHandler(check_delay)
+
+    async def initialize(self) -> None:
+        self.subscription_id = str(uuid.uuid4())
+        await self.task_handler.initialize()
+        await self.pubsub.subscribe(
+            handler=self.handle_message,
+            key=const.PubSubChannels.RESOURCE_NOTIFY.format(kbid="*"),
+            group="materializer",
+            subscription_id=self.subscription_id,
+        )
+
+    async def finalize(self) -> None:
+        await self.pubsub.unsubscribe(self.subscription_id)
+        await self.task_handler.finalize()
+
+    async def handle_message(self, raw_data) -> None:
+        data = self.pubsub.parse(raw_data)
+        notification = writer_pb2.Notification()
+        notification.ParseFromString(data)
+
+        if notification.action != writer_pb2.Notification.Action.COMMIT:
+            return
+
+        self.task_handler.schedule(
+            notification.kbid, partial(self.process, notification.kbid)
+        )
```

Comment:
Maybe it's not the case but, could it be that we are scheduling this twice on resource creation?
Not sure if it's available but this could be scheduled only if the message comes from the writer.

Also deletions should schedule the materializer

Reply:
`writer_pb2.Notification.Action.COMMIT` action happens for both create and delete.

also, if many events happen at the same time, it's okay, it is done with a delayed handler and unique against the kb id.

The point of the implementation is to consolidate all potential changes into 1 action at some interval. For example, 100 writes in 30 seconds on one kb will only cause 1 re-calc to happen.

Reply:
my bad, I see what you mean now. I've updated it to only recalc on create/delete and not modified. That should do the trick.